File: Labs/Lab 7/Main.java
public int findTheCity(int n, int[][] edges, int distanceThreshold) {
    int[][] distances = new int[n][n];
    for (int i = 0; i < n; i++) {
        Arrays.fill(distances[i], 10001); // Large value for 'infinity'
        distances[i][i] = 0; // Distance to itself is 0
    }

    // Initial distances
    for (int[] edge : edges) {
        distances[edge[0]][edge[1]] = edge[2];
        distances[edge[1]][edge[0]] = edge[2]; // Bidirectional 
    }

    // Floyd-Warshall
    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                distances[i][j] = Math.min(distances[i][j], distances[i][k] + distances[k][j]);
            }
        }
    }

    int minCount = n; 
    int idealCity = 0;
    for (int i = 0; i < n; i++) {
        int count = 0;
        for (int j = 0; j < n; j++) {
            if (i != j && distances[i][j] <= distanceThreshold) {
                count++;
            }
        }
        if (count <= minCount) {
            minCount = count;
            idealCity = i;
        }
    }

    return idealCity;
}

File: Labs/Lab 7/Main.py
def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
    distances = [[float('inf')] * n for _ in range(n)]
    for i in range(n):
        distances[i][i] = 0

    for u, v, w in edges:
        distances[u][v] = distances[v][u] = w

    # Floyd-Warshall
    for k in range(n):
        for i in range(n):
            for j in range(n):
                distances[i][j] = min(distances[i][j], distances[i][k] + distances[k][j])

    min_count = n
    ideal_city = 0
    for i in range(n):
        count = sum(d <= distanceThreshold for d in distances[i] if d != 0)  # Exclude distance to self
        if count <= min_count:
            min_count = count
            ideal_city = i

    return ideal_city 
